proposals/storage-get-keys.md

```diff
@@ -0,0 +1,115 @@
+# Proposal: browser.storage.&lt;area>.getKeys
+
+**Summary**
+
+API to retrieve all keys for a given storage area in the `browser.storage` API.
+
+**Document Metadata**
+
+**Author:** [Oliver Dunk (Google)](https://github.com/oliverdunk)
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** [polywock](https://github.com/polywock)
+
+**Created:** 2024-06-28
+
+**Related Issues:** [#601](https://github.com/w3c/webextensions/issues/601)
+
+## Motivation
+
+### Objective
+
+This API enables developers to retrieve the keys in storage without needing to
+receive all of the data stored alongside those keys. Depending on the
+implementation, the impact of this optimization may differ - for example
+Chromium has only a limited ability to access data from storage without
+fetching the associated data. Regardless, this reduces the amount of data
+returned to the caller and allows for further optimizations in the future.
+
+#### Use Cases
+
+There are existing use cases for enumerating the data in extension storage. For
+example, extensions for Chrome that add extension storage areas to DevTools.
+While these also need a way to show the associated values, they may not want to
+load these upfront.
+
+Additionally, in some cases developers may want to access keys with a given
+prefix, for example if using a key like `foo:tabid`. This API proposal does not
+allow matching by prefix (see "Future Work") but developers would be able to do
+this themselves after obtaining the list of keys.
+
+### Known Consumers
+
+There is developer interest for this API in
+[#601](https://github.com/w3c/webextensions/issues/601). While we don't expect
+significant immediate usage, it is a small, well-scoped API and should not
+represent significant engineering effort.
+
+## Specification
+
+### Schema
+
+```ts
+interface StorageArea {
+    getKeys(): Promise<string[]>;
+}
+```
+
+The original issue proposed the `getAllKeys()` name, to match IndexedDB. This
+proposal suggests the slightly shorter `getKeys()`, since that seems equally
+clear and less verbose.
```

Comment:
I'm not against this name. Just think of many other APIs that use the name `keys()`, such as: `Object.keys()`, `Map.keys()`, `Set.keys()` `Array.keys()`, `URLSearchParams.keys()`, `CacheStorage.keys()`, `Cache.keys()`, `FormData.keys()`, `Headers.keys()`, `NodeList.keys()`.

Reply:
In your examples, `CacheStorage.keys()` and `Cache.keys()` return promises, the others are synchronous results (many being iterators).

The extensions API has historically used `get` prefixes due to the async nature of the APIs with callback functions. However, with our APIs now using promises, I think it does make sense to drop `get` in our APIs going forward.

Reply:
I think there's also an argument that we should continue with the existing naming convention until we can update the API surface as a whole. I don't feel too strongly though.

Reply:
I'm okay with `getAll` and `getAllKeys`. I'd rather not use `keys()`.

Reply:
Given there is a Chrome and Firefox preference for `getKeys()` or `getAllKeys()`, it seems like we should pick between those. @hanguokai / @polywock, do you have preferences between those? I would prefer `getKeys()` if we can all align, just to keep it short.

Reply:
I also prefer getKeys() as it's more flexible. The proposal's Future Work section mentions a potential filter parameter. getAllKeys() might not be compatible with such parameter. 

getKeys(filter)  // makes sense
getAllKeys(filter) // makes less sense

Reply:
The name `x.keys()` is widely used in the JavaScript ecosystem, whether synchronous or asynchronous methods. It is used much more than the name `x.getKeys()` or `x.getAllKeys()`.

Also, in the `browser.storage` namespace, it uses only the names `get()` and `set()`, rather than `getValues()` or `getItems()`. So `get` is not used as a prefix here. In other words, the get prefix is not an existing naming convention in this namespace.

So I still suggest using the name `keys()`. But it doesn't really matter what it's called.

Reply:
> get() and set(), rather than getValues() or getItems()

Many namespaces in the Web Extension API have a primary `get()`, like `browser.tabs.get()` or `browser.windows.get()`. Additional methods that retrieve values have a get prefix like `tabs.getCurrent()`, `tabs.getSelected`, etc. 

> So get is not used as a prefix here. In other words, the get prefix is not an existing naming convention in this namespace.

`storage.getBytesInUse()`

Note: I'm fine with either `keys()` or `getKeys()` (leaning towards `getKeys()`)

Reply:
I agree with @polywock that `getKeys()` makes more sense if `filter` is ever added in the future.

Reply:
Both names ("getKeys" or "keys") are fine, simple and clear, I have no objection to either.

In fact, developers may primarily use it as an intermediate method, and wrap it as:
```js
async function getValues(filter) {
    let targetKeys = (await browser.storage.local.getKeys()).filter(filter);
    return browser.storage.local.get(targetKeys);
}
```